src/EditorFeatures/TestUtilities/Threading/StaTaskScheduler.cs

```diff
@@ -55,45 +56,42 @@ public StaTaskScheduler()
         {
             using (var threadStartedEvent = new ManualResetEventSlim(initialState: false))
             {
-                DispatcherSynchronizationContext synchronizationContext = null;
                 StaThread = new Thread(() =>
                 {
-                    var oldContext = SynchronizationContext.Current;
-                    try
-                    {
-                        // All WPF Tests need a DispatcherSynchronizationContext and we dont want to block pending keyboard
-                        // or mouse input from the user. So use background priority which is a single level below user input.
-                        synchronizationContext = new DispatcherSynchronizationContext();
+                    // All WPF Tests need a DispatcherSynchronizationContext and we dont want to block pending keyboard
+                    // or mouse input from the user. So use background priority which is a single level below user input.
+                    var synchronizationContext = new DispatcherSynchronizationContext();
 
-                        // xUnit creates its own synchronization context and wraps any existing context so that messages are
-                        // still pumped as necessary. So we are safe setting it here, where we are not safe setting it in test.
-                        SynchronizationContext.SetSynchronizationContext(synchronizationContext);
+                    // xUnit creates its own synchronization context and wraps any existing context so that messages are
+                    // still pumped as necessary. So we are safe setting it here, where we are not safe setting it in test.
+                    SynchronizationContext.SetSynchronizationContext(synchronizationContext);
 
-                        threadStartedEvent.Set();
+                    threadStartedEvent.Set();
 
-                        Dispatcher.Run();
-                    }
-                    finally
+                    while (true)
                     {
-                        SynchronizationContext.SetSynchronizationContext(oldContext);
+                        if (!_messageQueue.TryTake(out var action))
+                            action = _messageQueue.Take(_cancellationSource.Token);
```

Comment:
What effect does the cancellation token here have? Is there some concern that once the token is triggered, that this will also end the thread prematurely?

Reply:
Put another way, when you're cancelling this in the same place we previously cancelled the Dispatcher.ExitAllFrames() -- in either case the thread will quit, right?